src/ember-app.js

```diff
@@ -242,12 +243,27 @@ class EmberApp {
    * @param {Object} bootOptions An object containing the boot options that are used by
    *                             by ember to decide whether it needs to do rendering or not.
    * @param {Object} result
+   * @param {Boolean} buildSandboxPerVisit if true, a new sandbox will
+   *                                       **always** be created, otherwise one
+   *                                       is created for the first request
+   *                                       only
    * @return {Promise<instance>} instance
    */
-  async visitRoute(path, fastbootInfo, bootOptions, result) {
-    let app = await this.buildApp();
-    result.applicationInstance = app;
-
+  async _visit(path, fastbootInfo, bootOptions, result, buildSandboxPerVisit) {
+    let shouldBuildApp = buildSandboxPerVisit || this._applicationInstance === undefined;
+
+    let app = shouldBuildApp ? await this.buildApp() : this._applicationInstance;
+
+    if (buildSandboxPerVisit) {
+      // entangle the specific application instance to the result, so it can be
+      // destroyed when result._destroy() is called (after the visit is
+      // completed)
+      result.applicationInstance = app;
+    } else {
+      // save the created application instance so that we can clean it up when
+      // this instance of `src/ember-app.js` is destroyed (e.g. reload)
+      this._applicationInstance = app;
+    }
```

Comment:
I think this is a bit confusing in terms of variable naming. IIUC, we are deciding here whether to build a new EmberApp every request or re-use the old one and still create a new EmberAppInstance. Please correct me if I am wrong.

If that is the case, shouldn't we call it `this._application` and `result.application`. It isn't an instance per say. I know the already existing code calls it `appInstance` and appInstanceInstance` but that can easily cause confusion :)

Reply:
> I think this is a bit confusing in terms of variable naming.

Yeah, I don't disagree, but we are talking about instances of `Ember.Application` **and** `Ember.ApplicationInstance`. We have to call them _something_...

>  It isn't an instance per say.

Well, it is actually. It is the result of `Ember.Application.create()`, it is an `Ember.Application` instance (hence the name `this._applicationInstance`). The other thing we get is an `Ember.ApplicationInstance` instance (and is called `this._applicationInstanceInstance`).

---

tldr; I erred on the side of "correct but overly verbose", I could change it to `.application` and `.instance` if you prefer.

Reply:
yeah my only concern was around accidentally misread of the variables due to repeated subwords.

Reply:
Yeah :( including `Instance` in the name of a class is bad news bears 😭

Reply:
why not use `app` as variable name?  I agree using `appInstance` as a variable name is a bit confusing while reading the code.

Reply:
“App” is significantly overused as well. Is it an instance of `src/ember-app.js` in this project? Is it an instance of `lib/broccoli/ember-app.js` in ember-cli? Is it an Ember.Application instance? Is it the conceptual “application” that we are rendering? 🙀 

Tbh, I’d rather make up completely fabricated names (that don’t use “instance” at all) and link to a naming doc to explain what each of the names mean.

Reply:
I agree, it's a tough problem.